### PR TITLE
fix: adding `Sanitize()` method to `Data` interface

### PR DESCRIPTION
**Pull request by 0xankit**
## Description

Closes: #164 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and please add links to any
relevant follow-up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
  in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (
  see [PR Targeting](https://github.com/AssetMantle/modules/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines
  for [building modules](https://github.com/AssetMantle/modules/blob/master/docs/building-modules)
- [ ] included the necessary unit and
  integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add your handle next to the items
reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the
  correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: schema/data/base/decData.go [hunk 1]
```diff
@@ -28,6 +28,9 @@ func (decData decData) Compare(listable traits.Listable) int {
 	if err != nil {
 		panic(err)
 	}
+	if decData.Value.Int == nil {
```
Comment (deepanshutr) [inline schema/data/base/decData.go:31]: do it through sanitize method

Comment (0xankit) [reply schema/data/base/decData.go]: done

File: schema/data/base/decData.go [hunk 2]
```diff
@@ -41,6 +48,9 @@ func (decData decData) String() string {
 	return decData.Value.String()
 }
 func (decData decData) Bytes() []byte {
+	if sanitizedDecData, err := decData.Sanitize(); err != nil {
+		decData.Value = sanitizedDecData.(data.DecData).Get()
```
Comment (0xankit) [inline schema/data/base/decData.go:52]: closes #163

File: schema/data/base/accAddressData.go [hunk 3]
```diff
@@ -30,6 +30,12 @@ func (accAddressData accAddressData) Compare(listable traits.Listable) int {
...
 		panic(err)
 	}
+	if sanitizedAccAddressData, err := accAddressData.Sanitize(); err != nil {
+		accAddressData.Value = sanitizedAccAddressData.(data.AccAddressData).Get()
+	}
+	if sanitizedCompareAccAddressData, err := compareAccAddressData.Sanitize(); err != nil {
+		compareAccAddressData.Value = sanitizedCompareAccAddressData.(data.AccAddressData).Get()
+	}
 
 	return bytes.Compare(accAddressData.Value.Bytes(), compareAccAddressData.Value.Bytes())
```
Comment (deepanshutr) [inline schema/data/base/accAddressData.go:40]: compare should be done on sanitized fields

File: schema/data/base/accAddressData.go [hunk 4]
```diff
@@ -57,6 +63,13 @@ func (accAddressData accAddressData) Get() sdkTypes.AccAddress {
 	return accAddressData.Value
 }
 
+func (accAddressData accAddressData) Sanitize() (data.Data, error) {
+	if accAddressData.Value.Empty() {
```
Comment (deepanshutr) [inline schema/data/base/accAddressData.go:67]: what if value is nil, will Empty() panic

File: schema/data/base/booleanData.go [hunk 5]
```diff
@@ -65,6 +65,10 @@ func (booleanData booleanData) Get() bool {
 	return booleanData.Value
 }
 
+func (booleanData booleanData) Sanitize() (data.Data, error) {
+	return booleanData, nil
```
Comment (deepanshutr) [inline schema/data/base/booleanData.go:69]: check if value in boolean data is nil

File: schema/data/base/decData.go [hunk 6]
```diff
@@ -29,6 +29,13 @@ func (decData decData) Compare(listable traits.Listable) int {
...
 	}
 
+	if sanitizedDecData, err := decData.Sanitize(); err != nil {
+		decData.Value = sanitizedDecData.(data.DecData).Get()
+	}
+	if sanitizedDecData, err := compareDecData.Sanitize(); err != nil {
+		compareDecData.Value = sanitizedDecData.(data.DecData).Get()
+	}
+
 	if decData.Value.GT(compareDecData.Value) {
```
Comment (deepanshutr) [inline schema/data/base/decData.go:39]: use sanitized value to check

File: schema/data/base/accAddressData.go (hunk 4 above)
Comment (0xankit) [reply schema/data/base/accAddressData.go:66]: Empty() method checks for both
```
func (aa AccAddress) Empty() bool {
	if aa == nil {
		return true
	}

	aa2 := AccAddress{}
	return bytes.Equal(aa.Bytes(), aa2.Bytes())
}
```
unit test for same
https://github.com/AssetMantle/modules/blob/b5012c0df64fdc6baf40cb40755f40718dc4edc1/schema/data/base/accAddressData_test.go#L89-L92

File: schema/data/base/decData.go [hunk 7]
```diff
@@ -60,6 +70,13 @@ func (decData decData) Get() sdkTypes.Dec {
 	return decData.Value
 }
 
+func (decData decData) Sanitize() (data.Data, error) {
+	if decData.Value.Int == nil {
```
Comment (muku314115) [inline schema/data/base/decData.go:74]: It appears that there is no reference for Int in my cosmos-sdk version, can you please test this, The only option I can find is BigInt()

**Review by 0xankit — COMMENTED** [review]
@deepanshutr need your review in this PR

File: schema/data/base/decData.go (hunk 7 above)
Comment (0xankit) [reply schema/data/base/decData.go:74]: Int is being used as anonymous field here which is of the type big.Int. BigInt() is a method but we needed to check the value here

Comment (0xankit) [reply schema/data/base/decData.go:74]: Here is relevant unit tests:
https://github.com/AssetMantle/modules/blob/3eb675145202b5e88e8f4d0e4afa0725e9686a61/schema/data/base/decData_test.go#L113-L118

**Pull request closed without merging**